hazelcast-jet-sql/src/main/java/com/hazelcast/jet/sql/impl/JetSqlBackend.java

```diff
@@ -228,17 +235,49 @@ private ExecutionPlan toPlan(RelNode rel, List<String> fieldNames, OptimizerCont
         boolean isStreaming = containsStreamSource(rel);
         boolean isInsert = physicalRel instanceof TableModify;
 
+        List<Permission> permissions = extractPermissions(physicalRel);
         if (isInsert) {
             DAG dag = createDag(physicalRel);
-            return new ExecutionPlan(dag, isStreaming, true, null, null, planExecutor);
+            return new ExecutionPlan(dag, isStreaming, true, null, null, planExecutor, permissions);
         } else {
             QueryId queryId = QueryId.create(nodeEngine.getLocalMember().getUuid());
             DAG dag = createDag(new JetRootRel(physicalRel, nodeEngine.getThisAddress(), queryId));
             SqlRowMetadata rowMetadata = createRowMetadata(fieldNames, physicalRel.schema().getTypes());
-            return new ExecutionPlan(dag, isStreaming, false, queryId, rowMetadata, planExecutor);
+            return new ExecutionPlan(dag, isStreaming, false, queryId, rowMetadata, planExecutor, permissions);
         }
     }
 
+    private List<Permission> extractPermissions(PhysicalRel physicalRel) {
+        List<Permission> permissions = new ArrayList<>();
+
+        physicalRel.accept(new RelShuttleImpl() {
+            @Override
+            public RelNode visit(TableScan scan) {
+                addPermissionForTable(scan.getTable(), ActionConstants.ACTION_READ);
+                return super.visit(scan);
+            }
+
+            @Override
+            public RelNode visit(RelNode other) {
```

Comment:
Can't we use `visit(LogicalTableModify modify)`? Maybe we could skip the `table != null` condition in `addPermissionForTable`.

Reply:
The `InsertPhysicalRel` doesn't extend `LogicalTableModify`, but only `TableModify`. I actually don't know why `RelShuttle` has the `Logical` subclasses for almost everything except for `TableScan`...